source-genesys/source_genesys/api.py

```diff
@@ -0,0 +1,185 @@
+
+    API docs - https://developer.genesys.cloud/routing/conversations/conversations-apis#post-api-v2-analytics-conversations-details-jobs
+    """
+    # Submit job.
+    url = f"{COMMON_API}.{domain}/api/v2/analytics/conversations/details/jobs"
+    body = {
+        "interval": f"{_dt_to_str(start_date)}/{_dt_to_str(end_date)}",
+    }
+
+    response = CreateJobResponse.model_validate_json(
+        await http.request(log, url, method="POST", json=body)
+    )
+
+    job_id = response.jobId
+
+    # Query job status.
+    url = f"{COMMON_API}.{domain}/api/v2/analytics/conversations/details/jobs/{job_id}"
+
+    state = "QUEUED"
+    rate_limiter = RateLimiter()
+    # 3 seconds anecdotally seems like the minimum time for a job to complete.
+    rate_limiter.delay = 3.0 
+    while state != "FULFILLED":
+        delay = rate_limiter.delay
+        await asyncio.sleep(delay)
+
+        response = CheckJobStatusResponse.model_validate_json(
+            await http.request(log, url)
+        )
+
+        state = response.state
+
+        match state:
+            case "FAILED":
+                raise RuntimeError("Conversations job failed. Please inspect the error and update job details.", response.errorMessage)
+            case "CANCELLED" | "EXPIRED":
+                raise RuntimeError(f"Conversations job status is {state}. Job request body was: ", body)
+
+        rate_limiter.update(delay, True)
+
+    # Paginate through results.
+    url = f"{COMMON_API}.{domain}/api/v2/analytics/conversations/details/jobs/{job_id}/results"
+    params: dict[str, str | int] = {
+        "pageSize": 1000,
+    }
+
+    while True:
+        response = JobResultsResponse.model_validate_json(
+            await http.request(log, url, params=params)
+        )
+
+        conversations = response.conversations
+        for conversation in conversations:
+            yield conversation
+
+        cursor = response.cursor
+        if not cursor:
+            break
+
+        params.update({
+            "cursor": cursor,
+        })
+
+
+async def backfill_conversations(
+        http: HTTPSession,
+        domain: str,
+        log: Logger,
+        page_cursor: PageCursor | None,
+        cutoff: LogCursor,
+) -> AsyncGenerator[Conversation | PageCursor, None]:
+    """
+    Submits an asynchronous job for all conversations from the page_cursor (config start date) and the cutoff.
+    """
+    assert isinstance(page_cursor, str)
+    assert isinstance(cutoff, datetime)
+
+    start_date = _str_to_dt(page_cursor)
+
+    # Since cutoff is shifted backwards 2 days, it's possible for the cutoff to be before the start date.
+    # In this situation, incremental replication will cover everything that should be backfilled, and the
+    # backfill can safely return.
+    if start_date > cutoff:
+        return
+
+    async for result in _perform_conversation_job(
+        http, domain, log, start_date, end_date=cutoff,
+    ):
+        yield result
```

Comment:
No, we don't need to have both backfill & incremental parts since the backfill & incremental parts are essentially the same minus their start date. We can definitely just have the incremental part start at the config's start date & continue until the present. I'll rewrite it to do that.

I'll address the checkpointing question in your other comment instead of duplicating here.